nova/virt/vmwareapi/vmops.py

```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
+
+        dest = target.prepare_ds_transfer(context, "PUT", dest_ds_ref,
+                                          get_path.rel_path)
+        ds_url = dest['ds_url']
+        ticket = dest['ticket']
+
+        if cdrom.connectable.connected:
+            self._force_disconnect_cdrom(instance, cdrom)
+
+        with contextlib.closing(self._get_ds_file_handle(
+                get_path, ds_ref, method='GET')) as get:
+            put = rw_handles.FileWriteHandle(ds_url,
+                    cookies=ticket,
+                    file_size=get.get_size())
+            with contextlib.closing(put):
+                shutil.copyfileobj(get, put)
+
+        ds_url = ds_obj.DatastoreURL.urlparse(ds_url)
+        ds_path = ds_obj.DatastorePath(ds_url.datastore_name, ds_url.path)
+        return ds_path
+
+    def _force_disconnect_cdrom(self, instance, cdrom):
+        vm_ref = vm_util.get_vm_ref(self._session, instance)
+        cf = self._session.vim.client.factory
+
+        backing = cf.create("ns0:VirtualCdromRemoteAtapiBackingInfo")
+        backing.deviceName = ""
+        backing.useAutoDetect = True
+
+        device_change = vm_util.create_virtual_cdrom_spec(cf, None,
+                cdrom.controllerKey, None,
+                cdrom.unitNumber, cdrom.key, backing)
+
+        config_spec = cf.create('ns0:VirtualMachineConfigSpec')
+        config_spec.deviceChange = [device_change]
+
+        reconfig_task = self._session._call_method(self._session.vim,
+                                                   "ReconfigVM_Task", vm_ref,
+                                                   spec=config_spec)
+        task_completed = threading.Event()
+
+        def set_task_completed(gt):
+            task_completed.set()
+
+        wait_for_task = utils.spawn(
+                self._session.wait_for_task, reconfig_task)
+        wait_for_task.link(set_task_completed)
+
+        while not task_completed.is_set():
+            question = self._session._call_method(vutil,
+                    "get_object_property", vm_ref, "summary.runtime.question")
+            if question and any(message
+                                for message in question.message
+                                if message.id == "msg.cdromdisconnect.locked"):
+                yes = next(info
+                           for info in question.choice.choiceInfo
+                           if info.label == "button.yes")
+                LOG.warning("Forcing disconnect for cdrom", instance=instance)
+                self._session._call_method(self._session.vim, "AnswerVM",
+                                           vm_ref,
+                                           questionId=question.id,
+                                           answerChoice=yes.key)
+            task_completed.wait(1)
+
+    def reconfigure_vm_device_change(self, context, instance, devices):
+        if not devices:
+            return
+        serialized = [vutil.serialize_object(spec) for spec in devices]
```

Comment:
Why do we have to serialize the specs if we don't push them over the wire but instead call `vm_util` to apply them to the VC?

Reply:
Just to get a more compact logging. It becomes obsolete with sapcc/oslo.vmware#26.
The monkey-patch calls the serializer with a different argument, which is more extensive

Reply:
Ah, I'm just not reading things right. We just serialize for logging and then don't use the `serialized` anymore. I somehow saw it used in the `vm_util.reconfigure_vm_device_change()` call below O.o